// @electron-forge/publisher-github/dist/util/github.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const plugin_retry_1 = require("@octokit/plugin-retry");
const rest_1 = require("@octokit/rest");
const debug_1 = __importDefault(require("debug"));
const logInfo = (0, debug_1.default)('electron-forge:publisher:github:info');
const logDebug = (0, debug_1.default)('electron-forge:publisher:github:debug');
class GitHub {
    constructor(authToken = undefined, requireAuth = false, options = {}) {
        const noOp = () => {
            /* Intentionally does nothing */
        };
        this.options = {
            ...options,
            log: {
                debug: logDebug.enabled ? logDebug : noOp,
                error: console.error,
                info: logInfo.enabled ? logInfo : noOp,
                warn: console.warn,
            },
            userAgent: 'Electron Forge',
        };
        if (authToken) {
            this.token = authToken;
        }
        else if (process.env.GITHUB_TOKEN) {
            this.token = process.env.GITHUB_TOKEN;
        }
        else if (requireAuth) {
            throw new Error('Please set GITHUB_TOKEN in your environment to access these features');
        }
    }
    getGitHub() {
        const options = { ...this.options };
        if (this.token) {
            options.auth = this.token;
        }
        const RetryableOctokit = rest_1.Octokit.plugin(plugin_retry_1.retry);
        const github = new RetryableOctokit(options);
        return github;
    }
    // Based on https://github.com/cli/cli/blob/b07f955c23fb54c400b169d39255569e240b324e/pkg/cmd/release/upload/upload.go#L131-L153
    static sanitizeName(name) {
        return (node_path_1.default
            .basename(name)
            // Remove diacritics (e.g. é -> e)
            .normalize('NFD')
            .replace(/\p{Diacritic}/gu, '')
            // Replace special characters with dot
            .replace(/[^\w_.@+-]+/g, '.')
            // Replace multiple dots with a single dot
            .replace(/\.+/g, '.')
            // Remove leading dot if present
            .replace(/^\./g, '')
            // Remove trailing dot if present
            .replace(/\.$/g, ''));
    }
}
exports.default = GitHub;
