// @electron-forge/publisher-github/dist/PublisherGithub.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublisherGithub = void 0;
const node_path_1 = __importDefault(require("node:path"));
const publisher_base_1 = require("@electron-forge/publisher-base");
const request_error_1 = require("@octokit/request-error");
const chalk_1 = __importDefault(require("chalk"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const log_symbols_1 = __importDefault(require("log-symbols"));
const mime_types_1 = __importDefault(require("mime-types"));
const github_1 = __importDefault(require("./util/github"));
const no_release_error_1 = __importDefault(require("./util/no-release-error"));
class PublisherGithub extends publisher_base_1.PublisherBase {
    constructor() {
        super(...arguments);
        this.name = 'github';
    }
    async publish({ makeResults, setStatusLine, }) {
        const { config } = this;
        const perReleaseArtifacts = {};
        for (const makeResult of makeResults) {
            const release = makeResult.packageJSON.version;
            if (!perReleaseArtifacts[release]) {
                perReleaseArtifacts[release] = [];
            }
            perReleaseArtifacts[release].push(makeResult);
        }
        if (!(config.repository &&
            typeof config.repository === 'object' &&
            config.repository.owner &&
            config.repository.name)) {
            throw new Error('In order to publish to GitHub, you must set the "repository.owner" and "repository.name" properties in your Forge config. See the docs for more info');
        }
        const github = new github_1.default(config.authToken, true, config.octokitOptions);
        github.getGitHub();
        for (const releaseVersion of Object.keys(perReleaseArtifacts)) {
            let release;
            const artifacts = perReleaseArtifacts[releaseVersion];
            const releaseName = `${config.tagPrefix ?? 'v'}${releaseVersion}`;
            setStatusLine(`Searching for target release: ${releaseName}`);
            try {
                release = (await github.getGitHub().repos.listReleases({
                    owner: config.repository.owner,
                    repo: config.repository.name,
                    per_page: 100,
                })).data.find((testRelease) => testRelease.tag_name === releaseName);
                if (!release) {
                    throw new no_release_error_1.default(404);
                }
            }
            catch (err) {
                if (err instanceof no_release_error_1.default && err.code === 404) {
                    // Release does not exist, let's make it
                    release = (await github.getGitHub().repos.createRelease({
                        owner: config.repository.owner,
                        repo: config.repository.name,
                        tag_name: releaseName,
                        name: releaseName,
                        draft: config.draft !== false,
                        prerelease: config.prerelease === true,
                        generate_release_notes: config.generateReleaseNotes === true,
                    })).data;
                }
                else {
                    // Unknown error
                    throw err;
                }
            }
            let uploaded = 0;
            const updateUploadStatus = () => {
                setStatusLine(`Uploading distributable (${uploaded}/${artifacts.length} to ${releaseName})`);
            };
            updateUploadStatus();
            await Promise.all(artifacts
                .flatMap((artifact) => artifact.artifacts)
                .map(async (artifactPath) => {
                const done = () => {
                    uploaded += 1;
                    updateUploadStatus();
                };
                const artifactName = node_path_1.default.basename(artifactPath);
                const sanitizedArtifactName = github_1.default.sanitizeName(artifactName);
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const asset = release.assets.find((item) => item.name === sanitizedArtifactName);
                if (asset !== undefined) {
                    if (config.force === true) {
                        await github.getGitHub().repos.deleteReleaseAsset({
                            owner: config.repository.owner,
                            repo: config.repository.name,
                            asset_id: asset.id,
                        });
                    }
                    else {
                        return done();
                    }
                }
                try {
                    const { data: uploadedAsset } = await github
                        .getGitHub()
                        .repos.uploadReleaseAsset({
                        owner: config.repository.owner,
                        repo: config.repository.name,
                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                        release_id: release.id,
                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                        url: release.upload_url,
                        // https://github.com/octokit/rest.js/issues/1645
                        data: (await fs_extra_1.default.readFile(artifactPath)),
                        headers: {
                            'content-type': mime_types_1.default.lookup(artifactPath) || 'application/octet-stream',
                            'content-length': (await fs_extra_1.default.stat(artifactPath)).size,
                        },
                        name: artifactName,
                    });
                    if (uploadedAsset.name !== sanitizedArtifactName) {
                        // There's definitely a bug with GitHub.sanitizeName
                        console.warn(log_symbols_1.default.warning, chalk_1.default.yellow(`Expected artifact's name to be '${sanitizedArtifactName}' - got '${uploadedAsset.name}'`));
                    }
                }
                catch (err) {
                    // If an asset with that name already exists, it's either a bug with GitHub.sanitizeName
                    // where it did not sanitize the artifact name in the same way as GitHub did, or there
                    // was simply a race condition with uploading artifacts with the same name
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    if (err instanceof request_error_1.RequestError &&
                        err.status === 422 &&
                        err.response?.data?.errors?.[0].code ===
                            'already_exists') {
                        console.error(`Asset with name '${artifactName}' already exists - there may be a bug with Forge's GitHub.sanitizeName util`);
                    }
                    throw err;
                }
                return done();
            }));
        }
    }
}
exports.default = PublisherGithub;
exports.PublisherGithub = PublisherGithub;
